DOI: https://www.nature.com/articles/s41586-020-2871-y

New genome assemblies have been arriving at a rapidly increasing pace, thanks to decreases in sequencing costs and improvements in third-generation sequencing technologies[1](https://www.nature.com/articles/s41586-020-2871-y#ref-CR1 "Eid, J. et al. Real-time DNA sequencing from single polymerase molecules. Science 323, 133–138 (2009)."),[2](https://www.nature.com/articles/s41586-020-2871-y#ref-CR2 "Weisenfeld, N. I., Kumar, V., Shah, P., Church, D. M. & Jaffe, D. B. Direct determination of diploid genome sequences. Genome Res. 27, 757–767 (2017)."),[3](https://www.nature.com/articles/s41586-020-2871-y#ref-CR3 "Jain, M., Olsen, H. E., Paten, B. & Akeson, M. The Oxford Nanopore MinION: delivery of nanopore sequencing to the genomics community. Genome Biol. 17, 239 (2016)."). For example, the number of vertebrate genome assemblies currently in the NCBI (National Center for Biotechnology Information) database[4](https://www.nature.com/articles/s41586-020-2871-y#ref-CR4 "Kitts, P. A. et al. Assembly: a resource for assembled genomes at NCBI. Nucleic Acids Res. 44 (D1), D73–D80 (2016).") increased by more than 50% to 1,485 assemblies in the year from July 2018 to July 2019. In addition to this influx of assemblies from different species, new human de novo assemblies[5](https://www.nature.com/articles/s41586-020-2871-y#ref-CR5 "Jain, M. et al. Nanopore sequencing and assembly of a human genome with ultra-long reads. Nat. Biotechnol. 36, 338–345 (2018).") are being produced, which enable the analysis of not only small polymorphisms, but also complex, large-scale structural differences between human individuals and haplotypes. This coming era and its unprecedented amount of data offer the opportunity to uncover many insights into genome evolution but also present challenges in how to adapt current analysis methods to meet the increased scale. Cactus[6](https://www.nature.com/articles/s41586-020-2871-y#ref-CR6 "Paten, B. et al. Cactus: algorithms for genome multiple sequence alignment. Genome Res. 21, 1512–1528 (2011)."), a reference-free multiple genome alignment program, has been shown to be highly accurate, but the existing implementation scales poorly with increasing numbers of genomes, and struggles in regions of highly duplicated sequences. Here we describe progressive extensions to Cactus to create Progressive Cactus, which enables the reference-free alignment of tens to thousands of large vertebrate genomes while maintaining high alignment quality. We describe results from an alignment of more than 600 amniote genomes, which is to our knowledge the largest multiple vertebrate genome alignment created so far.